### PR TITLE
[QMS-317] Enable user to add POI to project via right-click

### PR DESCRIPTION
**Pull request by pingurus**
_**Note: Do not delete any of the sections**_
_**Answer them all. Replace the descriptive text**_
_**by your answer**_

**What is the linked issue for this pull request (start with a `#`):** QMS-#317

**Describe roughly what you have done:**

I added the possibility to add POIs via right-click to a project.

**What steps have to be done to perform a simple smoke test:**

1. Open poi file and load pois
2. right-click POIs
3. See it is added with name and description

**Does the code comply to the coding rules and naming conventions [Coding Guidelines](https://github.com/Maproom/qmapshack/wiki/DeveloperCodingGuideline):**

- [x] yes

**Is every user facing string in a tr() macro?**

- [x] yes

**Did you add the ticket number and title into the changelog? Keep the numeric order in each release block.**

- [x] yes, I didn't forget to change changelog.txt

### CODE REVIEW

File: src/qmapshack/poi/CPoiDraw.cpp [hunk 1]
```diff
@@ -137,6 +138,36 @@ void CPoiDraw::loadPoiPath(QSettings &cfg)
     poiPaths = cfg.value("poiPaths", poiPaths).toStringList();
 }
 
+poi_t CPoiDraw::findPOICloseBy(const QPoint &px) const
+{
+    poi_t poi;
+    CPoiItem::mutexActivePois.lock();
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiDraw.cpp:144]: Use `QMutexLocker`. Much safer.

File: src/qmapshack/poi/CPoiDraw.cpp [hunk 2]
```diff
@@ -137,6 +138,36 @@ void CPoiDraw::loadPoiPath(QSettings &cfg)
...
+    if(poiList)
+    {
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiDraw.cpp:146]: With `QMutexLocker` you could simplify the code:

```
if(!poiList)
{
    return poi;
}
``

File: src/qmapshack/poi/CPoiDraw.cpp [hunk 3]
```diff
@@ -137,6 +138,36 @@ void CPoiDraw::loadPoiPath(QSettings &cfg)
...
+        for(int i = 0; i < poiList->count(); i++)
+        {
+            auto item = poiList->item(i);
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiDraw.cpp:149]: I do not really like auto. Sometimes it's really nice however IDEs can't resolve `auto`. This makes the code hard to read. Please stick with explicit types as long as possible.

File: src/qmapshack/poi/CPoiDraw.cpp [hunk 4]
```diff
@@ -137,6 +138,36 @@ void CPoiDraw::loadPoiPath(QSettings &cfg)
...
+
+            if(!item || item->getPoifile().isNull())
+            {
+                // as all active maps have to be at the top of the list
+                // it is ok to break as soon as the first map with no
+                // active files is hit.
+                break;
+            }
+
+            item->getPoifile()->findPOICloseBy(px, poi);
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiDraw.cpp:159]: The only API exposed by item should be the one of `CPoiItem`. At this point I do not care if there is a poi file or what ever. I call `item->findPoiCloseBy()` and let the item do what ever it has to do.

File: src/qmapshack/poi/CPoiDraw.cpp [hunk 5]
```diff
@@ -137,6 +138,36 @@ void CPoiDraw::loadPoiPath(QSettings &cfg)
...
+            if(poi.pos != NOPOINTF)
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiDraw.cpp:160]: `findPOICloseBy()` could have a boolean return value to tell if the search has been successful.

File: src/qmapshack/poi/CPoiPOI.cpp [hunk 6]
```diff
@@ -126,6 +127,43 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
...
+        {
+            QPointF pt = poiItemFound.coordinates;
+            poi->convertRad2Px(pt);
+            QPointF x = px - pt;
+            if(x.manhattanLength() < 10)
+            {
+                poiItem.pos = poiItemFound.coordinates;
+                for(auto tag : {"name=", "brand=", "operator="})
+                {
+                    auto matches = poiItemFound.data.filter(tag);
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiPOI.cpp:149]: again `auto` is a bad choice. Everyone has to lookup the type of data and the return type of filter to understand that auto is in fact a `QStringList`. It should be a `const QStringList&` and I am not sure if auto does that.

File: src/qmapshack/poi/CPoiPOI.cpp [hunk 7]
```diff
@@ -126,6 +127,43 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
...
+                    if (matches.length() > 0)
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiPOI.cpp:150]: well I would use `isEmpty()` as it is what you want to know. But ok, that's a matter of taste.

**Review by kiozen — CHANGES_REQUESTED** [review]
My main concern with this MR is the fact that it is blocking the GUI if the POI thread is busy querying the database. And as it is busy each time I zoom/move the map the user will run into the problem of a blocking GUI quite frequently. 

Either you 
optimize the database access to happen in fractions of a second 

or 

you find a way to unlock the mutex on a very regular, frequent base

or 

you find another solution (e.g. do not search for POIs if the mutex would block)

**Comment by pingurus** [comment]
> My main concern with this MR is the fact that it is blocking the GUI if the POI thread is busy querying the database. And as it is busy each time I zoom/move the map the user will run into the problem of a blocking GUI quite frequently.
> 
> Either you
> optimize the database access to happen in fractions of a second
> 
> or
> 
> you find a way to unlock the mutex on a very regular, frequent base
> 
> or
> 
> you find another solution (e.g. do not search for POIs if the mutex would block)

I'm not sure if I understand you correctly, but nothing in this PR triggers access to the database. There is only a lookup in the table of loaded POIs.

The blocking of the user interface already occurs during the loading of the POIs to be displayed. This should either have been fixed in #302 (well, to late..), or when tackling #307. If you prefer, I can first work on #307 before proceeding with the tasks that emerged from #308 / #309.

**Comment by kiozen** [comment]
So far the POI loading takes a long time but since it is running in it's own thread and nothing competes it just takes time but does not block the GUI.

Now you introduce the POI lookup locking `mutexActivePois` which is totally ok. However while the POI thread loads POIs in the background this mutex will block until the mutex is unlocked by the POI thread. As the method is called in the main thread - the GUI - thread the complete GUI is blocked. In an event driven GUI application all methods called must not block for a longer timer. 

As this PR introduces the problem it has to be solved by this PR or we prioritize another PR to solve it first.

**Comment by pingurus** [comment]
I've made some benchmarking on the current feature/pois branch for the `CPOIPoi::draw()` method. The results for loading all POIs from the water category in the Baden-Wuerttemberg map (where I can see the effect described when using the branch of this PR) are the following: 
```
total Time: 8965 ms
locked Time: 3 ms
unlocked Time: 5996 ms
missing Time: 2966 ms
mean of unlockedTimes: 0.7305068226120858 ms
standard deviation of unlockedTimes: 1.7457090631014056 ms
min of unlockedTimes: 0 ms
max of unlockedTimes: 32 ms
```
For me this looks like the mutex is regularly unlocked while loading the POIs from the sql file. So I think the best solution to fix the hanging of the GUI is to do a `CPoiItem::mutexActivePois.tryLock();` and to skip checking for POIs if it can't lock the mutex.

File: src/qmapshack/poi/CPoiDraw.cpp (hunk 4 above)
Comment (pingurus) [reply src/qmapshack/poi/CPoiDraw.cpp:159]: Should I somehow restrict access to `CPOIPoi::findPOICloseBy` or is it enough to just provide a `CPoiItem::findPOIClosBy`?

**Comment by pingurus** [comment]
Loading of the POIs is still slow, but it no longer blocks the GUI.

File: src/qmapshack/poi/CPoiDraw.cpp (hunk 4 above)
Comment (kiozen) [reply src/qmapshack/poi/CPoiDraw.cpp]: Imho it's enough to provide a `CPoiItem::findPOIClosBy()`.

File: src/qmapshack/poi/CPoiPOI.cpp [hunk 8]
```diff
@@ -126,6 +127,45 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
...
+                {
+                    const QStringList& matches = poiItemFound.data.filter(tag);
+                    if (!matches.isEmpty())
+                    {
+                        poiItem.name = matches[0];
+                        poiItem.name.replace(tag, "");
+                        break;
+                    }
+                }
+                if(poiItem.name == "")
```
Comment (kiozen) [inline src/qmapshack/poi/CPoiPOI.cpp:157]: `isEmpty()` is always preferred over comparison with a null string. I would even expect it to be more performant.

**Review by kiozen — CHANGES_REQUESTED** [review]
GUI interaction is much better now. And I guess it's acceptable for the user not  to select POIs while rendering is still in progress. Hopefully we can optimize that.

Only two minor cosmetic nitpicks left. I would merge when they are done if that's ok for you.

**Pull request merged**